src/services/wallets/EthWalletConnector.ts

```diff
@@ -0,0 +1,150 @@
+  contractSigner?: ContractSigner;
+  connector: Connector;
+  config: Config;
+
+  constructor(config: Config) {
+    const connector = config.connectors.find((c) => c.name === 'MetaMask');
+
+    if (!connector) {
+      throw new MetamaskError('MetaMask connector not found.');
+    }
+
+    this.connector = connector;
+
+    const provider = {
+      getSigner: () => ({
+        signMessage: async (message: any) => {
+          const arg = message instanceof String ? message : { raw: message };
+
+          const ethAccount = getAccount(config);
+
+          return await signMessage(config, {
+            message: arg as any,
+            account: ethAccount.address,
+            connector: this.connector,
+          });
+        },
+      }),
+    };
+    const signer = new InjectedEthereumSigner(provider as any);
+    signer.setPublicKey = async () => {
+      const message = 'Sign this message to connect to Network Portal';
+      const ethAccount = getAccount(config);
+
+      const signature = await signMessage(config, {
+        message: message,
+        account: ethAccount.address,
+        connector: this.connector,
+      });
+      const hash = await hashMessage(message);
+      const recoveredKey = await recoverPublicKey({
+        hash,
+        signature,
+      });
+      signer.publicKey = Buffer.from(toBytes(recoveredKey));
+    };
+
+    const aoSigner: AoSigner = async ({ data, tags, target }) => {
+      const ethAccount = getAccount(config);
+
+      if (!signer.publicKey) {
+        await signer.setPublicKey();
+      }
+
+      const ETHEREUM_MAINNET_CHAIN_ID = 1;
+
+      if (ethAccount.chainId !== ETHEREUM_MAINNET_CHAIN_ID) {
+        try {
+          await this.connector.switchChain?.({
+            chainId: ETHEREUM_MAINNET_CHAIN_ID,
+          });
+        } catch (error) {
+          // Handle different switch errors
+          if (error instanceof UserRejectedRequestError) {
+            throw new Error('User rejected network switch');
+          } else if (error instanceof SwitchChainError) {
+            throw new Error('Unable to switch to Ethereum Mainnet');
+          } else {
+            throw new Error('Unexpected error switching networks');
+          }
+        }
+      }
+
+      const dataToSign = typeof data === 'string' ? data : new Uint8Array(data);
+      const dataItem = createData(dataToSign, signer, {
+        tags,
+        target,
+        anchor: Math.round(Date.now() / 1000)
+          .toString()
+          .padStart(32, Math.floor(Math.random() * 10).toString()),
+      });
+
+      const res = await dataItem.sign(signer).then(async () => ({
+        id: dataItem.id,
+        raw: dataItem.getRaw(),
+      }));
+      return res;
+    };
+
+    this.config = config;
+    this.contractSigner = aoSigner;
+  }
+
+  async connect(): Promise<void> {
+    try {
+      localStorage.setItem('walletType', WALLET_TYPES.ETHEREUM);
```

Comment:
Wondering if we should specify ETHEREUM_MAINNET

Reply:
for the wallet_types? Not sure it's quite right since it's really metamask we're connecting. Or are you talking about forcing a switch to mainnet when connecting? also not sure we need that until we really need it (such as in the signing flow).